crates/runtime/README.md

```diff
@@ -0,0 +1,662 @@
+  - External HTTP requests (fetch)
+
+*[Read more...](#host-functions-1)*
+
+### Resource limits
+
+The `VMLimits` struct defines constraints on various resources:
+
+  - Memory usage
+  - Stack size
+  - Number and size of registers
+  - Logging and event limits
+  - Storage key and value size limits
+
+### Storage
+
+The runtime includes an abstract `Storage` trait and an in-memory implementation
+(`InMemoryStorage`) for persisting state.
+
+*[Read more...](#storage-1)*
+
+### Error handling
+
+A comprehensive error handling system is implemented, covering various
+scenarios:
+
+  - Host errors
+  - Wasm execution errors
+  - Storage errors
+  - Function call errors
+
+*[Read more...](#error-handling-1)*
+
+### Serialisation and deserialisation
+
+The runtime uses [Serde][] for JSON serialisation and deserialisation of
+input/output data.
+
+## Operational flow
+
+  1. Load Wasm module
+  2. Set up execution environment (memory, limits, storage)
+  3. Resolve and link host functions
+  4. Execute specified function in the Wasm module
+  5. Handle return values, logs, and events
+  6. Clean up and return execution outcome
+
+```mermaid
+graph TD
+    A[Load Wasm Module] --> B[Set up Execution Environment]
+    B --> C[Create Wasmer Engine and Store]
+    C --> D[Set Memory Limits and Tuning]
+    D --> E[Instantiate Wasm Module]
+    E --> F[Resolve and Link Host Functions]
+    F --> G[Execute Specified Function]
+    G --> H{Execution Successful?}
+    H -->|Yes| I[Collect Results]
+    H -->|No| J[Handle Errors]
+    I --> K[Process Logs and Events]
+    J --> K
+    K --> L[Clean Up Resources]
+    L --> M[Return Execution Outcome]
+
+    subgraph "Host Function Calls"
+        G --> N[Memory Operations]
+        G --> O[Register Operations]
+        G --> P[Storage Interactions]
+        G --> Q[Logging and Events]
+        G --> R[External Requests]
+    end
+
+    subgraph "Error Handling"
+        J --> S[VM Runtime Errors]
+        J --> T[Function Call Errors]
+        J --> U[Host Errors]
+        J --> V[Wasm Traps]
+    end
+```
+
+## Detailed component descriptions
+
+### Topology
+
+```mermaid
+graph TD
+    A[Calimero Runtime] --> B[VMLogic]
+    A --> C[Wasm Module]
+    A --> D[Resource Limits]
+    
+    B --> E[Storage]
+    B --> F[Memory Management]
+    B --> G[Registers]
+    B --> H[Context]
+    B --> I[Host Functions]
+    
+    I --> J[Panic Handling]
+    I --> K[Memory Operations]
+    I --> L[Register Operations]
+    I --> M[Storage Operations]
+    I --> N[Logging/Events]
+    I --> O[External Requests]
+    
+    C --> P[Guest Code]
+    P -.-> I
+    
+    E --> Q[InMemoryStorage]
+    E --> R[Other Storage Implementations]
+    
+    F --> S[Wasmer Memory]
+    
+    T[Error Handling] --> U[VMRuntimeError]
+    T --> V[FunctionCallError]
+    T --> W[HostError]
+    T --> X[WasmTrap]
+    
+    A --> T
+    
+    Y[Serialisation] --> Z[Input Deserialisation]
+    Y --> AA[Output Serialisation]
```

Comment:
Fair, I think that's more significance than I was intending - it's just a visual representation for topology rather than implying anything deeper. As a relationship diagram was desired, that's why I put one in 🙂 

I'll remove those entries based on the general points about (de)serialisation responsibility.

Reply:
As well as removing those entries I have also simplified the graph.